### ISSUE TITLE
XCM Runtime API fee estimation UX improvement

### ISSUE THREAD

**Issue by mrshiposha** [paritytech/polkadot-sdk#6126]
We (Unique Network's team) have been building a JS package to estimate XCM fees. Although it kind of works, the estimation process is very inconvinient. Moreover, we can only accurately estimate the execution fees. To estimate the delivery fees, we'd need to analyze each program and find the instructions that forward new programs. This requires constant maintenance of the library since, with new XCM versions, there could be new such instructions. It would be good to have a future-proof universal solution backed by a generic Runtime API. Also, new kinds of fees can be introduced in the future, which could require special handling similar to delivery fees (see [this discussion](https://github.com/polkadot-fellows/xcm-format/pull/53)).

The current implementation of dry-running doesn't help here since if there are not enough fees inside the dry-running extrinsic/xcm-program, it will just fail without telling us how much fee asset we'd need. The current design of calling `XcmPaymentApi` first and then `DryRunApi` has proven inconvenient.

## Suggestion

Make the XCM executor participate in the dry-running process directly instead of relying on the different parts of its configuration, such as its Routers.

The XCM executor could record all the taken fees directly (inside the `take_fee` function). It should not fail if there are not enough fees during the dry-running; it should just set a flag about that and report an error at the very end of the program execution.
>There is an issue with the `Trader`, though. It just returns an error if something goes wrong inside the `buy_weight`. It doesn't tell us the fee amount required. Also, there could be an error if a user provided a wrong fee asset, which can't be used to cover the execution fees.

If we make the XCM executor do this, we don't need to reimplement the fee calculation logic separately in the Runtime API. Instead, the `DryRunApi` should be able to read this info from the XCM executor after a program dry-running.

Also, the XCM executor should record the forwarded programs directly (inside the `send` function). This will eliminate the #5878 instantly.

`DryRunApi` should then collect all the info it currently provides and **also** report all the required fees that the XCM executor tried to collect. This way, we don't need to think about different kinds of fees. It's the XCM executor's business. Users want to ensure they pay enough for everything to get their XCM program through all the chains.

In the end, the `DryRunApi` would just take the needed info from the XCM executor. This means the `DryRunApi` would need exactly one implementation, and any chain team that already configured the XCM executor could just plug the `DryRunApi` in their Runtime.

If implemented, we can remove most of the `XcmPaymentApi` methods. We'd need to preserve the `queryAcceptablePaymentAssets` method only.

**Comment by acatangiu** [comment]
> The current implementation of dry-running doesn't help here since if there are not enough fees inside the dry-running extrinsic/xcm-program, it will just fail without telling us how much fee asset we'd need. 

The recommended flow is to dry-run with some large value for all assets used for fees anywhere along the path, get the exact values used for fees from the dry-run, add a small buffer to that (to account for any potential brewing congestion) and submit live transaction.

**Comment by acatangiu** [comment]
> Make the XCM executor participate in the dry-running process directly instead of relying on the different parts of its configuration, such as its Routers.

The XCM executor is just one component in a Runtimes's XCM processing pipeline. There are also barriers and routers/exporters (maybe others?). Any one of these components can and do mutate some XCM programs, so trying to handle it inside the XCM executor is a losing battle.

For example, bridges use "exporting" routers (e.g. [SovereignPaidRemoteExporter](https://github.com/paritytech/polkadot-sdk/blob/d4409e3bad65cde9097e73f9f7bfd602e442375c/polkadot/xcm/xcm-builder/src/universal_exports.rs#L263)) that "capture" outbound messages destined for other `GlobalConsensus`s **after** they leave the executor, wraps them in bridge-specific outer XCMs and routes them through BridgeHub (or wherever the bridge for that particular other `GlobalConsensus` is deployed).

For correctly estimating fees, you need to also account the "proxy" hops (like BridgeHub) and you need to account for potential mutations to an XCM program that happen _after_ they leave the XCM executor.

The current design of dry-running covers the full flow by injecting input XCMs at the very beginning of a runtime's XCM processing pipeline and capturing output XCMs at the very end of the pipeline (capturing exact bytes that are enqueued in outbound transport queues).

**Comment by mrshiposha** [comment]
I understand your point about the routers, barriers, etc. I agree that we need to capture the XCM programs at the very end of the pipeline, even after the routers do their job. Capturing them from the outbound queues will also eliminate #5878 while maintaining generality. It will also correctly reflect any mutations to the sent programs so that we will see the actual sent programs and their actual destinations.

----

>The recommended flow is to dry-run with some large value for all assets used for fees anywhere along the path, get the exact values used for fees from the dry-run, add a small buffer to that (to account for any potential brewing congestion) and submit live transaction.

If a user doesn't have a large amount of the fee asset, the extrinsic dry-run fails immediately. Namely, the `WithdrawAsset` will fail. It is a corner case, considering the usual fee amount (which is small), but this case is a valid one (for each possible fee asset). Also, it is unclear what this arbitrarily large fee asset amount for dry-running should be (again, for each possible fee asset).
That's an inconvenience from the API user perspective. They still have to guess what amount to put in to get even the fee estimation right.

Also, I find it inconvenient (when using the API) to handle different kinds of fees separately (execution and delivery fees). Learning **all** the fees from the dry-run result would be far better.
There is also a guess factor regarding the delivery fees: we can't estimate them before the dry run as we don't know what programs will be sent and where, and the dry run might fail if we don't have enough funds to cover the delivery fees.
(There are workarounds. Still, it would simplify the API usage greatly if we could learn the needed fees after the dry-run result).

There is also an inconvenience from the API implementor's perspective, as any implementor of the XCM Runtime APIs is forced to duplicate the execution fee calculation logic to return the required payment from the `query_weight_to_asset_fee`. The XCM executor already knows how to take fees for the XCM programs (except for the Trader case, as I wrote in the issue description). Why not make it expose the fee info? Reimplementing the fee calculation logic could lead to divergence between fees reported by the API and actually taken during the execution.
I'm assuming that only the XCM executor takes the fees (except for the "pre-paid" execution where the weight_credit != 0, but this usually means that the execution is performed as part of an extrinsic).

**Comment by acatangiu** [comment]
> If a user doesn't have a large amount of the fee asset, the extrinsic dry-run fails immediately. Namely, the WithdrawAsset will fail. It is a corner case, considering the usual fee amount (which is small), but this case is a valid one (for each possible fee asset). Also, it is unclear what this arbitrarily large fee asset amount for dry-running should be (again, for each possible fee asset).
That's an inconvenience from the API user perspective. They still have to guess what amount to put in to get even the fee estimation right.

I think the disconnect comes from the perception that we should design chain for direct end-user interactions. This is a losing strategy. Chains are not end-user products, they are infrastructure for other end-user products.

End users should not be using this API, nor should they be sending manually crafted XCMs. The user interacts with a dapp's UI; given a user's intention the Wallets/UIs/dApps job is to build the relevant XCM and among other things, also present to the user some estimated fees. For estimating the fees the wallet/UI can dry-run in the background, even multiple times if necessary, to "compile" the final result. In dry-runs it can use any or _all_ of the funds in the user's account, it can dry run with another/test account, it can do a lot of things and it's all offchain automated logic.

I really don't see a "better solution" to do this by moving some of this complexity on-chain...

> There is also an inconvenience from the API implementor's perspective, as any implementor of the XCM Runtime APIs is forced to duplicate the execution fee calculation logic to return the required payment from the query_weight_to_asset_fee. The XCM executor already knows how to take fees for the XCM programs (except for the Trader case, as I wrote in the issue description). Why not make it expose the fee info? Reimplementing the fee calculation logic could lead to divergence between fees reported by the API and actually taken during the execution.

This looks like a good improvement to do 👍 cc @franciscoaguirre

**Comment by mrshiposha** [comment]
>End users should not be using this API

I formulated the thought poorly, sorry. By an "API user," I meant Wallet/dApps/Libraries/UI developers, **not** an end-user. Specifically, the fee estimation code could've been more straightforward in our JS library.

>In dry-runs it can use any or *all* of the funds in the user's account

If we could learn the balance of any asset in a standard way on every chain in the ecosystem, it would be a good solution.
I like the idea of dry-running with a fee value equal to the user's total amount. Can we add a new method to the `XcmPaymentApi`, say, `balance_of(AccountId, AssetId)`, reporting the user's balance in the selected currency?

(I still have a feeling that IF precisely one entity on a chain takes XCM fees, we could expose that info in the dry-run result. But I agree it can be challenging. And maybe even impossible if several entities are at play taking fees).

**Comment by acatangiu** [comment]
> Yet, it is located within the common crate for Asset Hubs. Can we move it to the rest of the XCM Runtime APIs (polkadot/xcm/xcm-runtime-apis)?

Yeah, I knew there was one but didn't find it with a quick search, probably because I was also looking in the wrong place.
We can move it, yes.

cc @x3c41a 

---

Worth mentioning that there are other, pallet-specific runtime APIs useful to XCM usecases, which will continue living under their pallets, rather than moving them to `polkadot/xcm/xcm-runtime-apis`.

A quick way to see all available ones (for a particular chain) is looking at the live chain: e.g. https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fstatemint.public.curie.radiumblock.co%2Fws#/runtime

**Comment by mrshiposha** [comment]
However, there is a catch when using all the user's funds for dry-running: existential deposits. If the user's account has no additional providers, the dry run with all user funds will fail since the transaction in question would kill the account.
